src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/CodeBasedDependencyAlgorithm.cs

```diff
@@ -118,23 +120,77 @@ public static void AddDependenciesDueToMethodCodePresence(ref DependencyList dep
             }
         }
 
-        private static void AddPInvokeParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        public static void AddInterestingInteropConstructedTypeDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
         {
-            if (parameter.IsDelegate)
+            if (type.IsDelegate)
             {
-                dependencies.Add(factory.NecessaryTypeSymbol(parameter), "Delegate Marshalling Stub");
+                var delegateType = type as MetadataType;
+                if (delegateType != null && delegateType.HasCustomAttribute("System.Runtime.InteropServices", "UnmanagedFunctionPointerAttribute"))
+                {
+                    AddDependenciesDuePInvokeDelegate(ref dependencies, factory, delegateType);
+                }
+            }
+            else if (type.IsValueType && type.IsTypeDefinition && !(type is NativeStructType))
+            {
+                var structType = type as MetadataType;
+                if (structType != null && structType.HasCustomAttribute("System.Runtime.InteropServices", "StructLayoutAttribute"))
```

Comment:
Do you see this path to be ever taken? StructLayoutAttribute is synthetic attribute. I do not think it will be ever in metadata.